Source/DCS/NET/doc/man/man3/SockCOM.3
.TH "SockCOM" 3 "Mon Mar 26 2012" "Version 1.0" "NET" \" -*- nroff -*-
.ad l
.nh
.SH NAME
SockCOM \- 
.SH SYNOPSIS
.br
.PP
.PP
\fC#include <SockCOM\&.h>\fP
.PP
Inherits \fBSockIO\fP\&.
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBSockCOM\fP (const std::string &comPort, unsigned int baudRate=SOCKCOM_BR9600, unsigned int dataBits=SOCKCOM_BITS8, unsigned int parity=SOCKCOM_PARITYNONE, unsigned int stopBits=SOCKCOM_STOPBITS1, unsigned int flowControl=SOCKCOM_FCNONE)"
.br
.RI "\fICreate a new \fBSockCOM\fP object bound to the specified device\&. \fP"
.ti -1c
.RI "\fBSockCOM\fP (unsigned int comPort=1, unsigned int baudRate=SOCKCOM_BR9600, unsigned int dataBits=SOCKCOM_BITS8, unsigned int parity=SOCKCOM_PARITYNONE, unsigned int stopBits=SOCKCOM_STOPBITS1, unsigned int flowControl=SOCKCOM_FCNONE)"
.br
.RI "\fICreate a new \fBSockCOM\fP object bound to the specific port\&. \fP"
.ti -1c
.RI "virtual \fB~SockCOM\fP ()"
.br
.RI "\fIVirtual destructor\&. \fP"
.ti -1c
.RI "SockCOMIOHandle \fBgetDescriptor\fP () const "
.br
.RI "\fIRetrieve handle to serial device\&. \fP"
.ti -1c
.RI "std::string \fBgetCOMPort\fP () const "
.br
.RI "\fIRetrieve name of serial device\&. \fP"
.ti -1c
.RI "unsigned int \fBgetBaudRate\fP () const "
.br
.RI "\fIRetrieve baud rate setting for the serial device\&. \fP"
.ti -1c
.RI "unsigned int \fBgetDataBits\fP () const "
.br
.RI "\fIRetrieve number of data bits setting for the serial device\&. \fP"
.ti -1c
.RI "unsigned int \fBgetParity\fP () const "
.br
.RI "\fIRetrieve parity setting for the serial device\&. \fP"
.ti -1c
.RI "unsigned int \fBgetStopBits\fP () const "
.br
.RI "\fIRetrieve number of stop bits setting for the serial device\&. \fP"
.ti -1c
.RI "unsigned int \fBgetFlowControl\fP () const "
.br
.RI "\fIRetrieve flow control setting for the serial device\&. \fP"
.ti -1c
.RI "virtual bool \fBisValid\fP () const "
.br
.RI "\fIIndicates the handle references a valid (open) serial device\&. \fP"
.ti -1c
.RI "bool \fBisBlockingIO\fP () const "
.br
.RI "\fIIndicates blocking I/O mode\&. \fP"
.ti -1c
.RI "bool \fBsetBlockingIO\fP (bool block)"
.br
.RI "\fISet blocking I/O mode\&. \fP"
.ti -1c
.RI "int32_t \fBgetLastError\fP () const "
.br
.RI "\fIGet system error code\&. \fP"
.ti -1c
.RI "void \fBsetLastError\fP (int32_t error)"
.br
.RI "\fISet the system error code\&. \fP"
.ti -1c
.RI "virtual bool \fBopen\fP ()"
.br
.RI "\fIOpen serial device\&. \fP"
.ti -1c
.RI "virtual bool \fBclose\fP ()"
.br
.RI "\fIClose serial device\&. \fP"
.ti -1c
.RI "virtual bool \fBreadReady\fP (uint32_t sec=0, uint32_t usec=0) const "
.br
.RI "\fIIndicates data is available for reading\&. \fP"
.ti -1c
.RI "virtual bool \fBwriteReady\fP (uint32_t sec=0, uint32_t usec=0) const "
.br
.RI "\fIIndicates space is available for writing\&. \fP"
.ti -1c
.RI "virtual int32_t \fBread\fP (void *buffer, int32_t size, int32_t flags=0) const "
.br
.RI "\fIRead data\&. \fP"
.ti -1c
.RI "virtual int32_t \fBwrite\fP (const void *buffer, int32_t size, int32_t flags=0) const "
.br
.RI "\fIWrite data\&. \fP"
.in -1c
.SS "Protected Attributes"

.in +1c
.ti -1c
.RI "std::string \fBcomPort_\fP"
.br
.RI "\fICurrent COM port\&. \fP"
.ti -1c
.RI "unsigned int \fBbaudRate_\fP"
.br
.RI "\fIBaud rate\&. \fP"
.ti -1c
.RI "unsigned int \fBdataBits_\fP"
.br
.RI "\fIData bits\&. \fP"
.ti -1c
.RI "unsigned int \fBparity_\fP"
.br
.RI "\fIParity\&. \fP"
.ti -1c
.RI "unsigned int \fBstopBits_\fP"
.br
.RI "\fIStop bits\&. \fP"
.ti -1c
.RI "unsigned int \fBflowControl_\fP"
.br
.RI "\fIFlow control\&. \fP"
.ti -1c
.RI "SockCOMIOHandle \fBioHandle_\fP"
.br
.RI "\fIHandle to the current port\&. \fP"
.ti -1c
.RI "int \fBcurrentState_\fP"
.br
.RI "\fICurrent state of the object\&. \fP"
.ti -1c
.RI "int \fBlastError_\fP"
.br
.RI "\fILast error reported\&. \fP"
.in -1c
.SH "Detailed Description"
.PP 
Cross-platform library for reading and writing to COM ports\&. Can read and write to COM 1-n\&. 
.SH "Constructor & Destructor Documentation"
.PP 
.SS "\fBSockCOM::SockCOM\fP (const std::string &comPort, unsigned intbaudRate = \fCSOCKCOM_BR9600\fP, unsigned intdataBits = \fCSOCKCOM_BITS8\fP, unsigned intparity = \fCSOCKCOM_PARITYNONE\fP, unsigned intstopBits = \fCSOCKCOM_STOPBITS1\fP, unsigned intflowControl = \fCSOCKCOM_FCNONE\fP)"
.PP
Create a new \fBSockCOM\fP object bound to the specified device\&. Create a new \fBSockCOM\fP object bound to the specified device\&. Intended for use with RS-232 devices with names that do not match the common pattern for an operating system, such as USB to serial adapter cables which are assigned names such as '/dev/ttyUSB0' on Linux systems 
.PP
References open()\&.
.SS "\fBSockCOM::SockCOM\fP (unsigned intcomPort = \fC1\fP, unsigned intbaudRate = \fCSOCKCOM_BR9600\fP, unsigned intdataBits = \fCSOCKCOM_BITS8\fP, unsigned intparity = \fCSOCKCOM_PARITYNONE\fP, unsigned intstopBits = \fCSOCKCOM_STOPBITS1\fP, unsigned intflowControl = \fCSOCKCOM_FCNONE\fP)"
.PP
Create a new \fBSockCOM\fP object bound to the specific port\&. Create a new \fBSockCOM\fP object bound to the specific port\&. The device name is determind by appending the specified port number to the following strings: 'COM' + comPort for the Microsoft Windows operating system '/dev/ttyS' + comPort for Linux systems '/dev/term/' + (comPort + 'a') for SunOS systems '/dev/ttyf' + comPort for other UNIX systems 
.PP
References comPort_, and open()\&.
.SH "Member Function Documentation"
.PP 
.SS "bool \fBSockCOM::close\fP ()\fC [virtual]\fP"
.PP
Close serial device\&. Close an open serial device\&. 
.PP
\fBReturns:\fP
.RS 4
a boolean value indicating the operation was successful if true, was unsuccessful if false\&. 
.RE
.PP

.PP
Implements \fBSockIO\fP\&.
.PP
References currentState_, ioHandle_, and isValid()\&.
.PP
Referenced by open(), and ~SockCOM()\&.
.SS "unsigned int \fBSockCOM::getBaudRate\fP () const"
.PP
Retrieve baud rate setting for the serial device\&. Returns the baud rate associated with the serial device\&. 
.PP
\fBReturns:\fP
.RS 4
an unsigned integer specifying the baud rate selected for the serial device\&. 
.RE
.PP

.PP
References baudRate_\&.
.SS "std::string \fBSockCOM::getCOMPort\fP () const"
.PP
Retrieve name of serial device\&. Returns the name of the associated serial device\&. 
.PP
\fBReturns:\fP
.RS 4
a string containing an operating system specific name of the associated serial device\&. 
.RE
.PP

.PP
References comPort_\&.
.SS "unsigned int \fBSockCOM::getDataBits\fP () const"
.PP
Retrieve number of data bits setting for the serial device\&. Returns the number of data bits per byte associated with the serial device\&. 
.PP
\fBReturns:\fP
.RS 4
an unsigned integer specifying the number of data bits per byte specified for the serial device\&. 
.RE
.PP

.PP
References dataBits_\&.
.SS "SockCOMIOHandle \fBSockCOM::getDescriptor\fP () const"
.PP
Retrieve handle to serial device\&. Returns the file handle for the serial device\&. 
.PP
\fBReturns:\fP
.RS 4
an operating system specific handle to the file descriptor associated with the serial device, or 0 if the device is closed\&. 
.RE
.PP

.PP
References ioHandle_\&.
.SS "unsigned int \fBSockCOM::getFlowControl\fP () const"
.PP
Retrieve flow control setting for the serial device\&. Returns the clow control setting associated with the serial device\&. 
.PP
\fBReturns:\fP
.RS 4
an unsigned integer specifying the flow control setting selected for the serial device\&. 
.RE
.PP

.PP
References flowControl_\&.
.SS "int32_t \fBSockCOM::getLastError\fP () const"
.PP
Get system error code\&. Returns the last known system error value\&. This function is not specific to any one serial device\&. The error value retrieved can be for any socket that is part of the overall application\&. 
.PP
\fBReturns:\fP
.RS 4
an error code indicating the reason for serial device failure\&. 
.RE
.PP

.PP
References lastError_\&.
.PP
Referenced by open()\&.
.SS "unsigned int \fBSockCOM::getParity\fP () const"
.PP
Retrieve parity setting for the serial device\&. Returns the parity setting associated with the serial device\&. 
.PP
\fBReturns:\fP
.RS 4
an unsigned integer specifying the parity setting selected for the serial device\&. 
.RE
.PP

.PP
References parity_\&.
.SS "unsigned int \fBSockCOM::getStopBits\fP () const"
.PP
Retrieve number of stop bits setting for the serial device\&. Returns the number of stop bits associated with the serial device\&. 
.PP
\fBReturns:\fP
.RS 4
an unsigned integer specifying the number of stop bits specified for the serial device\&. 
.RE
.PP

.PP
References stopBits_\&.
.SS "bool \fBSockCOM::isBlockingIO\fP () const\fC [virtual]\fP"
.PP
Indicates blocking I/O mode\&. Indicates that the serial device is or is not using blocking I/O\&. 
.PP
\fBReturns:\fP
.RS 4
a boolean value indicating that the serial device is using blocking I/O if true, is not using blocking I/O if false\&. 
.RE
.PP

.PP
Implements \fBSockIO\fP\&.
.PP
References ioHandle_\&.
.SS "bool \fBSockCOM::isValid\fP () const\fC [virtual]\fP"
.PP
Indicates the handle references a valid (open) serial device\&. Indicates that the serial object's handle is valid (open)\&. 
.PP
\fBReturns:\fP
.RS 4
a boolean value indicating the serial device handle is valid if true, and invalid if false\&. 
.RE
.PP

.PP
Reimplemented from \fBSockIO\fP\&.
.PP
References ioHandle_\&.
.PP
Referenced by close(), read(), readReady(), write(), writeReady(), and ~SockCOM()\&.
.SS "bool \fBSockCOM::open\fP ()\fC [virtual]\fP"
.PP
Open serial device\&. Opens the serial device for communication\&. 
.PP
\fBReturns:\fP
.RS 4
a boolean value indicating the operation was successful if true, was unsuccessful if false\&. 
.RE
.PP

.PP
Implements \fBSockIO\fP\&.
.PP
References baudRate_, close(), comPort_, currentState_, dataBits_, flowControl_, getLastError(), ioHandle_, parity_, setLastError(), and stopBits_\&.
.PP
Referenced by SockCOM()\&.
.SS "int32_t \fBSockCOM::read\fP (void *buffer, int32_tsize, int32_tflags = \fC0\fP) const\fC [virtual]\fP"
.PP
Read data\&. Reads data that is available in the serial device's input buffer\&. 
.PP
\fBParameters:\fP
.RS 4
\fIbuffer\fP a pointer to an object to be filled with the data in the socket buffer\&. 
.br
\fIsize\fP an integer specifying the size, in bytes, of 'buffer'\&. 
.br
\fIflags\fP an integer specifying the bit flags to be passed to network sockets when reading\&. Unused by serial devices\&. 
.RE
.PP
\fBReturns:\fP
.RS 4
the total number of bytes read\&. 
.RE
.PP

.PP
Implements \fBSockIO\fP\&.
.PP
References ioHandle_, and isValid()\&.
.SS "bool \fBSockCOM::readReady\fP (uint32_tsec = \fC0\fP, uint32_tusec = \fC0\fP) const\fC [virtual]\fP"
.PP
Indicates data is available for reading\&. Indicates that data is avialable in the seral device's buffer for reading\&. 
.PP
\fBParameters:\fP
.RS 4
\fIsec\fP an unsigned integer specifying the number of seconds to wait before returning\&. 
.br
\fIusec\fP an unsinged integer specifying the number of microseconds to wait before returning\&. 
.RE
.PP
\fBReturns:\fP
.RS 4
a boolean value indicating that data is available if true, no data is available if false\&. 
.RE
.PP

.PP
Implements \fBSockIO\fP\&.
.PP
References ioHandle_, and isValid()\&.
.SS "bool \fBSockCOM::setBlockingIO\fP (boolblock)\fC [virtual]\fP"
.PP
Set blocking I/O mode\&. Sets the serial device's I/O mode to blocking or non-blocking\&. 
.PP
\fBParameters:\fP
.RS 4
\fIblock\fP a boolean value indicating that the serial device should use blocking I/O if true, non-blocking I/O if false\&. 
.RE
.PP
\fBReturns:\fP
.RS 4
a boolean value indicating that I/O mode was successfully changed if true, was not successfully changed if false\&. 
.RE
.PP

.PP
Implements \fBSockIO\fP\&.
.PP
References ioHandle_\&.
.SS "void \fBSockCOM::setLastError\fP (int32_terror)"
.PP
Set the system error code\&. Sets the serial device's error code to the specified value\&. All serial devices share the same instance of the error code\&. 
.PP
\fBParameters:\fP
.RS 4
\fIerror\fP an integer specifying the value to which the error value is to be set\&. 
.RE
.PP

.PP
References lastError_\&.
.PP
Referenced by open()\&.
.SS "int32_t \fBSockCOM::write\fP (const void *buffer, int32_tsize, int32_tflags = \fC0\fP) const\fC [virtual]\fP"
.PP
Write data\&. Writes data to the network\&. 
.PP
\fBParameters:\fP
.RS 4
\fIbuffer\fP a pointer to an object filled with the data to be written\&. 
.br
\fIsize\fP an integer specifying the size, in bytes, of 'buffer'\&. 
.br
\fIflags\fP an integer specifying the bit flags to be passed to network sockets when writing\&. Unused by serial devices\&. 
.RE
.PP
\fBReturns:\fP
.RS 4
the total number of bytes written\&. 
.RE
.PP

.PP
Implements \fBSockIO\fP\&.
.PP
References ioHandle_, and isValid()\&.
.SS "bool \fBSockCOM::writeReady\fP (uint32_tsec = \fC0\fP, uint32_tusec = \fC0\fP) const\fC [virtual]\fP"
.PP
Indicates space is available for writing\&. Indicates that space in the serial device's buffer is avialable for writing\&. 
.PP
\fBParameters:\fP
.RS 4
\fIsec\fP an unsigned integer specifying the number of seconds to wait before returning\&. 
.br
\fIusec\fP an unsinged integer specifying the number of milliseconds to wait before returning\&. 
.RE
.PP
\fBReturns:\fP
.RS 4
a boolean value indicating that space is available if true, no space is available if false\&. 
.RE
.PP

.PP
Implements \fBSockIO\fP\&.
.PP
References ioHandle_, and isValid()\&.

.SH "Author"
.PP 
Generated automatically by Doxygen for NET from the source code\&.


Source/DCS/NET/src/SockCOM.cpp
// -*- mode: c++ -*-
/****************************************************************************
 *****                                                                  *****
 *****                   Classification: UNCLASSIFIED                   *****
 *****                    Classified By:                                *****
 *****                    Declassify On:                                *****
 *****                                                                  *****
 ****************************************************************************
 *
 *
 * Developed by: Naval Research Laboratory, Tactical Electronic Warfare Div.
 *               EW Modeling & Simulation, Code 5770
 *               4555 Overlook Ave.
 *               Washington, D.C. 20375-5339
 *
 * For more information please send email to simdis@enews.nrl.navy.mil
 *
 * July 6, 2004 - U.S. Naval Research Laboratory.
 *
 * The U.S. Government retains all rights to use, duplicate, distribute,
 * disclose, or release this software.
 ****************************************************************************
 *
 *
 */
#ifndef WIN32
#include <unistd.h>
#include <termios.h>
#else
#include <WinSock2.h>
#include <windows.h>
#endif

#include <stdio.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <string.h>
#include <errno.h>
#include "timec.h"
#include "iostreamc"
#include "sstreamc"

#include "sockdefs.h"
#include "SockIO.h"
#include "SockCOM.h"

// Define the default port names
#ifndef WIN32
# ifdef Linux
#  define DEVICE_STRING  "/dev/ttyS"
# else
#  ifdef Solaris
#   define DEVICE_STRING  "/dev/term/"
#  else
#   define DEVICE_STRING  "/dev/ttyf"
#  endif
# endif
#else
#  define DEVICE_STRING  "COM"
#endif

enum
{
  STATE_CLOSED=0,
  STATE_OPEN,
  STATE_ERROR
};

SockCOM::SockCOM (const std::string& comPort, unsigned int baudRate, unsigned int dataBits, unsigned int parity, unsigned int stopBits, unsigned int flowControl)
: SockIO(),
  comPort_(comPort),
  baudRate_(baudRate),
  dataBits_(dataBits),
  parity_(parity),
  stopBits_(stopBits),
  flowControl_(flowControl),
  ioHandle_(INVALID_HANDLE_VALUE),
  currentState_(STATE_CLOSED)
{
  open();
}

SockCOM::SockCOM (unsigned int comPort, unsigned int baudRate, unsigned int dataBits, unsigned int parity, unsigned int stopBits, unsigned int flowControl)
: SockIO(),
  baudRate_(baudRate),
  dataBits_(dataBits),
  parity_(parity),
  stopBits_(stopBits),
  flowControl_(flowControl),
  ioHandle_(INVALID_HANDLE_VALUE),
  currentState_(STATE_CLOSED)
{
  std::stringstream s;

#ifdef WIN32
  s << DEVICE_STRING << (int)(comPort);
#else
#ifdef Solaris
  s << DEVICE_STRING << ((int)(comPort) + (int)'a');
#else
  s << DEVICE_STRING << (int)(comPort);
#endif
#endif

  comPort_ = s.str();

  open();
}

SockCOM::~SockCOM()
{
  if (isValid())
  {
    close();
  }
}

SockCOMIOHandle SockCOM::getDescriptor() const
{
  return ioHandle_;
}

std::string SockCOM::getCOMPort() const
{
  return comPort_;
}

unsigned int SockCOM::getBaudRate() const
{
  return baudRate_;
}

unsigned int SockCOM::getDataBits() const
{
  return dataBits_;
}

unsigned int SockCOM::getParity() const
{
  return parity_;
}

unsigned int SockCOM::getStopBits() const
{
  return stopBits_;
}

unsigned int SockCOM::getFlowControl() const
{
  return flowControl_;
}

bool SockCOM::isValid() const
{
  return (ioHandle_ != INVALID_HANDLE_VALUE);
}

bool SockCOM::isBlockingIO() const
{
#ifndef WIN32
  // Unix...
  int flags;
  if ((flags = fcntl (ioHandle_, F_GETFL)) < 0)
    return true;
  return !(flags & O_NONBLOCK);
#else
  // Windows...
#endif

  return true;
}

bool SockCOM::setBlockingIO (bool block)
{
#ifndef WIN32
  // Unix...
  int flags;
  if ((flags = fcntl (ioHandle_, F_GETFL)) < 0)
    return false;
  if (block)
    flags = flags & ~O_NONBLOCK;
  else
    flags |= O_NONBLOCK;
  return (fcntl (ioHandle_, F_SETFL) >= 0);
#else
  // Windows...
  return false;
#endif
}

int32_t SockCOM::getLastError() const
{
  return lastError_;
}

void SockCOM::setLastError (int32_t error)
{
  lastError_ = error;
}

bool SockCOM::open()
{
  if (currentState_ != STATE_CLOSED)
    return false;

#ifndef WIN32
  ioHandle_ = ::open (comPort_.c_str(), O_RDWR | O_NOCTTY | O_NDELAY);
  if (ioHandle_ < 0)
  {
    ioHandle_ = INVALID_HANDLE_VALUE;
    setLastError (errno);
    if (getLastError() == EACCES)
      cerr<<"ERROR: current user needs read/write access on device "<< comPort_.c_str() <<endl;
    return false;
  }

  struct termios my_termios;
  tcgetattr(ioHandle_, &my_termios);

  // Modify settings for for raw data transmission (settings match those used by the pyserial Python module)
  my_termios.c_cflag |=  (CLOCAL|CREAD);
  my_termios.c_iflag &= ~(INLCR|IGNCR|ICRNL|IGNBRK|IUCLC|PARMRK);
  my_termios.c_oflag &= ~(OPOST);
  my_termios.c_lflag &= ~(ICANON|ECHO|ECHOE|ECHOK|ECHONL|ISIG|IEXTEN);

  // Poll and return what is already present with no wait
  my_termios.c_cc[VTIME] = 0;
  my_termios.c_cc[VMIN] = 0;

  // Set data bits
  my_termios.c_cflag &= ~(CSIZE);         // Clear all size bits
  switch(dataBits_)
  {
    case SOCKCOM_BITS5:
      my_termios.c_cflag |= CS5;
      break;
    case SOCKCOM_BITS6:
      my_termios.c_cflag |= CS6;
      break;
    case SOCKCOM_BITS7:
      my_termios.c_cflag |= CS7;
      break;
    default:
    case SOCKCOM_BITS8:
      my_termios.c_cflag |= CS8;
      break;
  }

  // Set parity - use non-POSIX CMSPAR for Linux and PAREXT for Solaris to enable mark/space parity
  // If CMSPAR was not defined in termios.h we will try to define it ourselves, although this may not
  // work for all Linux systems
  #if defined(Linux) && !defined(CMSPAR)
  #define CMSPAR 010000000000
  #endif
  my_termios.c_iflag &= ~(INPCK|ISTRIP);
  switch(parity_)
  {
  case SOCKCOM_PARITYSPACE:
    my_termios.c_cflag &= ~(PARODD);
#ifdef Linux
    my_termios.c_cflag |=  (PARENB|CMSPAR);
#else
    my_termios.c_cflag |=  (PARENB|PAREXT);
#endif
    break;
  case SOCKCOM_PARITYMARK:
#ifdef Linux
    my_termios.c_cflag |=  (PARENB|PARODD|CMSPAR);
#else
    my_termios.c_cflag |=  (PARENB|PARODD|PAREXT);
#endif
    break;
  case SOCKCOM_PARITYEVEN:
    my_termios.c_cflag &= ~(PARODD);
    my_termios.c_cflag |=  (PARENB);
    break;
  case SOCKCOM_PARITYODD:
    my_termios.c_cflag |= (PARENB|PARODD);
    break;
  default:
  case SOCKCOM_PARITYNONE:
    my_termios.c_cflag &= ~(PARENB|PARODD);
    break;
  }

  // Set stop bits
  switch(stopBits_)
  {
  default:
  case SOCKCOM_STOPBITS1:
    my_termios.c_cflag &= ~CSTOPB;
    break;
  case SOCKCOM_STOPBITS2:
    my_termios.c_cflag |=  CSTOPB;
    break;
  }

  // Intialize flow control settings (enable RTS and DTR lines)
  int lineFlags = TIOCM_DTR|TIOCM_RTS;
  ioctl(ioHandle_, TIOCMBIS, &lineFlags);
  my_termios.c_iflag &= ~(IXON|IXOFF|IXANY);
  my_termios.c_cflag |=  (CRTSCTS);

  // Set flow control
  if(flowControl_ == SOCKCOM_FCXONXOFF)
  {
    // Software flow control
    my_termios.c_iflag |=  (IXON|IXOFF);
  }
  else if(flowControl_ == SOCKCOM_FCRTSCTS)
  {
    // Clear RTS line enable bit
    lineFlags = TIOCM_RTS;
    ioctl(ioHandle_, TIOCMBIC, &lineFlags);
    my_termios.c_cflag |=  (CRTSCTS);
  }

  // Set baud rate
  switch(baudRate_)
  {
  case 115200:
    cfsetospeed(&my_termios, B115200);
    cfsetispeed(&my_termios, B115200);
    break;
  case 57600:
    cfsetospeed(&my_termios, B57600);
    cfsetispeed(&my_termios, B57600);
    break;
  case 38400:
    cfsetospeed(&my_termios, B38400);
    cfsetispeed(&my_termios, B38400);
    break;
  case 19200:
    cfsetospeed(&my_termios, B19200);
    cfsetispeed(&my_termios, B19200);
    break;
  case 9600:
    cfsetospeed(&my_termios, B9600);
    cfsetispeed(&my_termios, B9600);
    break;
  case 4800:
    cfsetospeed(&my_termios, B4800);
    cfsetispeed(&my_termios, B4800);
    break;
  case 2400:
    cfsetospeed(&my_termios, B2400);
    cfsetispeed(&my_termios, B2400);
    break;
  case 1200:
    cfsetospeed(&my_termios, B1200);
    cfsetispeed(&my_termios, B1200);
    break;
  case 600:
    cfsetospeed(&my_termios, B600);
    cfsetispeed(&my_termios, B600);
    break;
  case 300:
    cfsetospeed(&my_termios, B300);
    cfsetispeed(&my_termios, B300);
    break;
  case 110:
    cfsetospeed(&my_termios, B110);
    cfsetispeed(&my_termios, B110);
    break;
  default:
    cfsetospeed(&my_termios, B9600);
    cfsetispeed(&my_termios, B9600);
  }

  tcflush(ioHandle_, TCIFLUSH);
  tcsetattr(ioHandle_, TCSANOW, &my_termios);

#else
  std::string filename = std::string("\\\\.\\") + comPort_;
  ioHandle_ = CreateFile (filename.c_str(), GENERIC_READ|GENERIC_WRITE, 0, NULL, OPEN_EXISTING, 0, NULL);
  if (ioHandle_ == (HANDLE)HFILE_ERROR)
  {
    ioHandle_ = INVALID_HANDLE_VALUE;
    setLastError (errno);
    return false;
  }

  COMMTIMEOUTS commTimeout;
  if (!GetCommTimeouts (ioHandle_, &commTimeout))
  {
    setLastError (errno);
    close();
    return false;
  } // endif

  commTimeout.ReadIntervalTimeout = 100;
  commTimeout.ReadTotalTimeoutMultiplier = 0;
  commTimeout.ReadTotalTimeoutConstant = 100;
  commTimeout.WriteTotalTimeoutMultiplier = 100;
  commTimeout.WriteTotalTimeoutConstant = 50;

  if (!SetCommTimeouts (ioHandle_, &commTimeout))
  {
    setLastError (errno);
    close();
    return false;
  }

  DCB devicestate;
  memset(&devicestate, 0, sizeof(devicestate));
  devicestate.DCBlength = sizeof(devicestate);
  if (!GetCommState (ioHandle_, &devicestate))
  {
    setLastError (errno);
    close();
    return false;
  }

  devicestate.ByteSize = dataBits_;

  switch(parity_)
  {
  case SOCKCOM_PARITYSPACE:
    devicestate.fParity = TRUE;
    devicestate.Parity = SPACEPARITY;
    break;
  case SOCKCOM_PARITYMARK:
    devicestate.fParity = TRUE;
    devicestate.Parity = MARKPARITY;
    break;
  case SOCKCOM_PARITYEVEN:
    devicestate.fParity = TRUE;
    devicestate.Parity = EVENPARITY;
    break;
  case SOCKCOM_PARITYODD:
    devicestate.fParity = TRUE;
    devicestate.Parity = ODDPARITY;
    break;
  default:
  case SOCKCOM_PARITYNONE:
    devicestate.fParity = FALSE;
    devicestate.Parity = NOPARITY;
    break;
  }

  switch(stopBits_)
  {
  default:
  case SOCKCOM_STOPBITS1:
    devicestate.StopBits = ONESTOPBIT;
    break;
  case SOCKCOM_STOPBITS2:
    devicestate.StopBits = TWOSTOPBITS;
    break;
  }

  // Default to no flow control (RTS/DTR are always on)
  devicestate.fOutxDsrFlow = FALSE;
  devicestate.fDtrControl = DTR_CONTROL_ENABLE;
  devicestate.fOutxCtsFlow = FALSE;
  devicestate.fRtsControl = RTS_CONTROL_ENABLE;
  devicestate.fOutX = FALSE;
  devicestate.fInX = FALSE;

  if(flowControl_ == SOCKCOM_FCXONXOFF)
  {
    // Software flow control
    devicestate.fOutX = TRUE;
    devicestate.fInX = TRUE;
  }
  else if(flowControl_ == SOCKCOM_FCRTSCTS)
  {
    devicestate.fOutxCtsFlow = TRUE;
    devicestate.fRtsControl = RTS_CONTROL_HANDSHAKE;
  }

  switch(baudRate_)
  {
  case 256000:
    devicestate.BaudRate = CBR_256000;
    break;
  case 128000:
    devicestate.BaudRate = CBR_128000;
    break;
  case 115200:
    devicestate.BaudRate = CBR_115200;
    break;
  case 57600:
    devicestate.BaudRate = CBR_57600;
    break;
  case 56000:
    devicestate.BaudRate = CBR_56000;
    break;
  case 38400:
    devicestate.BaudRate = CBR_38400;
    break;
  case 19200:
    devicestate.BaudRate = CBR_19200;
    break;
  case 14400:
    devicestate.BaudRate = CBR_14400;
    break;
  case 9600:
    devicestate.BaudRate = CBR_9600;
    break;
  case 4800:
    devicestate.BaudRate = CBR_4800;
    break;
  case 2400:
    devicestate.BaudRate = CBR_2400;
    break;
  case 1200:
    devicestate.BaudRate = CBR_1200;
    break;
  case 600:
    devicestate.BaudRate = CBR_600;
    break;
  case 300:
    devicestate.BaudRate = CBR_300;
    break;
  case 110:
    devicestate.BaudRate = CBR_110;
    break;
  default:
    devicestate.BaudRate = CBR_9600;
  }

  if (!SetCommState (ioHandle_, &devicestate))
  {
    setLastError (errno);
    close();
    return false;
  }
#endif

  currentState_ = STATE_OPEN;
  return true;
}

bool SockCOM::close()
{
  if (isValid())
  {
#ifndef WIN32
    ::close (ioHandle_);
#else
    CloseHandle (ioHandle_);
#endif
    ioHandle_ = INVALID_HANDLE_VALUE;
  }

  currentState_ = STATE_CLOSED;
  return true;
}

bool SockCOM::readReady(uint32_t sec, uint32_t usec) const
{
  if (!isValid())
    return false;
#ifndef WIN32
  // UNIX reading
  fd_set fdSet;
  struct timeval timeout = {sec, usec};
  FD_ZERO (&fdSet);
  FD_SET (ioHandle_, &fdSet);
  int result = select (ioHandle_ + 1, &fdSet, NULL, NULL, &timeout);
  if (result == SOCKET_ERROR)
    return true;
  else if (result == 0)
    return false;

  return (FD_ISSET (ioHandle_, &fdSet)) ? true : false;

#else
  // Windows reading
  struct timeval tp;
  gettimeofday (&tp, 0);
  double endTime = tp.tv_sec + sec;
  double enduTime = tp.tv_usec + usec;
  while ((tp.tv_sec < endTime) || ((tp.tv_sec == endTime) && (tp.tv_usec <= enduTime)))
  {
    DWORD errorType;
    COMSTAT comStatus;
    if (ClearCommError (ioHandle_, &errorType, &comStatus))
    {
      if (comStatus.cbInQue > 0)
	return true;
    }
    else // Try to let the client pick up on the error
      return true;
    gettimeofday (&tp, 0);
  }
  return false;
#endif
}

bool SockCOM::writeReady(uint32_t sec, uint32_t usec) const
{
  if (!isValid())
    return false;
#ifndef WIN32
  // UNIX Writing
  fd_set writefd;
  struct timeval timeout={sec,usec};

  FD_ZERO (&writefd);
  FD_SET (ioHandle_, &writefd);
  int32_t result = select (ioHandle_ + 1, NULL, &writefd, NULL, &timeout);

  //For now return true so we try to read and get an error, rather than poll for ever on a bad socket
  if (result == SOCKET_ERROR)
    return true;
  else if (result == 0)
    return false;

  return (FD_ISSET (ioHandle_, &writefd)) ? true : false;
#else
  // Windows writing
  return true;
#endif
}

int32_t SockCOM::read(void* buffer, int32_t size, int32_t flags) const
{
  if (!isValid())
    return -1;

#ifndef WIN32
  // UNIX Writing
  return ::read (ioHandle_, buffer, size);
#else
  // Windows reading
  DWORD numRead = 0;
  if (!ReadFile (ioHandle_, buffer, size, &numRead, NULL))
  {
//     setLastError (errno);
    return -1;
  }
  return numRead;
#endif
}

int32_t SockCOM::write(const void* buffer, int32_t size, int32_t flags) const
{
  if (!isValid())
    return -1;

#ifndef WIN32
  // UNIX Writing
  return ::write (ioHandle_, buffer, size);
#else
  // Windows writing
  DWORD charsWritten;
  if (!WriteFile (ioHandle_, buffer, size, &charsWritten, NULL))
  {
//     setLastError (errno);
    return -1;
  }
  return charsWritten;
#endif

  return 0;
}
